Sources/SwiftParser/Statements.swift

```diff
@@ -877,11 +918,66 @@ extension Parser.Lookahead {
         // For example, could be the function call "discard()".
         return false
       }
+
+    case .then:
+      return isStartOfThenStatement(preferExpr: preferExpr)
+
     case nil:
       return false
     }
   }
 
+  /// Whether we're currently at a `then` token that should be parsed as a
+  /// `then` statement.
+  mutating func isStartOfThenStatement(preferExpr: Bool) -> Bool {
+    // If the feature is disabled, don't parse it.
+    guard self.experimentalFeatures.contains(.thenStatements) else {
+      return false
+    }
+    guard self.at(.keyword(.then)) else {
+      return false
+    }
+
+    // If we want to prefer an expr, and aren't at the start of a newline, then
+    // don't parse a ThenStmt.
+    if preferExpr && !self.atStartOfLine {
+      return false
+    }
+
+    let next = peek()
+
+    // If 'then' is followed by a binary or postfix operator, prefer to parse as
+    // an expr.
+    if BinaryOperatorLike(lexeme: next) != nil || PostfixOperatorLike(lexeme: next) != nil {
+      return false
+    }
+
+    switch next.rawTokenKind {
+    case .keyword:
+      // Treat 'is' and 'as' like the binary operator case, and parse as an
+      // expr.
+      switch Keyword(next.tokenText) {
+      case .as, .is:
+        return false
+      default:
+        break
+      }
```

Comment:
You should be able to simplify this a little bit by using `TokenSpec`

```suggestion
    switch PrepareForKeywordMatch(next) {
    case TokenSpec(.as), TokenSpec(.is):
      // Treat 'is' and 'as' like the binary operator case, and parse as an
      // expr.
      return false
```

Reply:
Cool!